WMS service: option to skip name attribute for groups in wms capabilities document

tomass opened the pull request:
## Description

When QGIS project contains layer groups, they are listed in WMS capabilities document alongside with layers. If a group (it's name as listed in capabilities) is included in WMS LAYERS alongside with names of layers in that group, QGIS would duplicate the layers: once for the group in LAYERS= and once for specific layer in LAYERS= Here is an example of result with duplicate layers - note the duplicate labels:
![image](https://github.com/qgis/QGIS/assets/969513/61f2a7af-50b9-4461-ac1e-08f011b9b596)

OGC WMS specification states that name attribute can be skipped. Technique of skipping name tag but leaving all other information including the title tag is used in say geoserver to avoid adding a group in LAYERS= but still being able to display the group in layer hierarchy.

## Change

I've added a project level setting "Skip name attribute for groups" which by default is not checked leaving the existing behaviour - groups do have a name attribute. This means functionality will not change for existing deployments.

If this setting is checked it will not add a name attribute for groups in WMS capabilities document.

Example excerpt from GetCapabilities response:

```
<Layer queryable="1">
  <Name>layer_name</Name> <--- this will be skipped when option is enabled
  <Title>Human readable name</Title>
```
## Our practice

We had a problem with duplicate layers when groups were included in LAYERS= of WMS request. Therefore we've been doing such change (removing of name attribute) manually from capabilities xml for several months with WMS clients using our services working as expected in QGIS as well as other on-line and desktop GIS clients.

tomass committed:
Option to skip name attribute for groups in wms capabilities document

tomass committed:
Option to skip name attribute for groups in wms capabilities document

tomass committed:
code style fix

tomass committed:
code style fix

tomass committed:
Update sip file

tomass committed:
Update sip file

tomass committed:
Merge remote-tracking branch 'upstream/master'

tomass committed:
Merge remote-tracking branch 'upstream/master'

tomass committed:
Merge remote-tracking branch 'upstream/master'

tomass committed:
Merge remote-tracking branch 'upstream/master'

tomass committed:
Merge remote-tracking branch 'upstream/master'

rldhont commented:
@tomass can you add a test please ?
Whithout a test, your addition could lost in the futur.

tomass commented:
> @tomass can you add a test please ? Whithout a test, your addition could lost in the futur.

  Yes, I can try doing that.
  Do you have any specific use case in mind? Or should I simply add two projects one with the new option switched on and one with it being off, and then test that the first one does not return a name tag in getCapabilities and the second one does return it?

tomass commented:
@rldhont I'm looking at both c++ and python test case scripts and so far cannot find a nice way to get capabilities output from two specific test projects I have. I will continue looking, but if you have a general picture how should I load a project and then get wms capabilities xml from that particular project, that would help a lot.

Gustry commented:
Can't you add one project [in the test data folder](https://github.com/qgis/QGIS/tree/master/tests/testdata/qgis_server) and add the corresponding test ?

tomass commented:
Thank you @Gustry , I've managed to find examples of loading the project from test data folders. But cannot find an example on how to get WMS GetFeature xml response from that project (I know how to do it in command line, but not in c++/python).

tomass commented:
Ok, test_qgsserver_wms_*.py seem to be doing what we need here.

tomass committed:
Added unit tests

tomass committed:
Added unit tests

tomass committed:
python lint

tomass committed:
python lint

tomass committed:
Fix class name

tomass committed:
Fix class name

tomass committed:
Fix ctest name

tomass committed:
Fix ctest name

tomass committed:
Fix spelling

tomass committed:
Fix spelling

tomass commented on tests/src/python/CMakeLists.txt:
```diff
@@ -586,4 +586,5 @@ if (WITH_SERVER)
   ADD_PYTHON_TEST(PyQgsServerRequest test_qgsserver_request.py)
   ADD_PYTHON_TEST(PyQgsServerResponse test_qgsserver_response.py)
   ADD_PYTHON_TEST(PyQgsServerConfigCache test_qgsserver_configcache.py)
+  ADD_PYTHON_TEST(PyQgsServerConfigCache test_qgsserver_wms_getcapabilities_group_name.py)
```
Fixed

Gustry commented on src/ui/qgsprojectpropertiesbase.ui:
```diff
@@ -2864,6 +2864,13 @@
+                     <item row="9" column="0">
+                      <widget class="QCheckBox" name="mSkipNameForGroupCheckBox">
+                       <property name="text">
+                        <string>Skip name attribute for groups</string>
```
Can you try to add a tooltip explaining the checkbox ? More or less inspired by the PR description ?

Because honestly, it can be difficult sometimes to understand some checkbox in this dialog, not your fault, just thinking about that just now ;-)

tomass replied:
No problem. Is the text "GetCapabilities response would skip name attribute for group items while leaving title attribute as is, thus showing the group but making it impossible to include the whole group (rather than it's elements) in subsequent GetMap request LAYERS parameter." clear enough and is it's length ok?

Gustry replied:
It's tricky to explain, I know :)
But at least, it's better than nothing, when we don't know the PR

tomass replied:
Fixed

tomass committed:
Added a tooltip

tomass committed:
Added a tooltip

tomass committed:
Merge remote-tracking branch 'upstream/master'

ptitjano committed:
qgsline3dsymbol: Rename height property to offset

The `height` property is in fact a vertical offset applied to the
line. Also, the `offset` phrase is already used by the terrain for
similar purposes.

height() are setHeight() are kept for API compatibility reasons.

ptitjano committed:
qgsline3dsymbolwidget: Rename height property to offset

See the previous commit.

ptitjano committed:
qgspolygon3dsymbol: Rename height property to offset

This is similar to the changes already applied to QgsLine3DSymbol.

height() are setHeight() are kept for API compatibility reasons.

ptitjano committed:
qgspolygon3dsymbolwidget: Rename height property to offset

See the previous commits.

ptitjano committed:
qgs3dutils: Rename height parameter in clampAltitudes to offset

See the previous commits.

tomass committed:
Merge remote-tracking branch 'upstream/master'

benoitdm-oslandia committed:
fix(cmake/FindPyQt6.py): remove acces to PyQt6.QtCore.PYQT_CONFIGURATION

PYQT_CONFIGURATION is only available on PyQt5

benoitdm-oslandia committed:
fix(cmake/FindQsci.cmake): set QSCI_VER to Qt major version instead of 5

As it did not work with qt6

nirvn committed:
Fix copy/pasting of features to/from the same layer ignores avoid overlap settings

nirvn committed:
Add test covering feature copy/pasting

Gustry reviewed (approved):
I have reviewed the UI&tests

elpaso committed:
Server: support both LAYER and LAYERS for describelayer

For consistency with other calls.

elpaso committed:
Fix aggregate function orderby

Fix #54580

nyalldawson committed:
Update src/app/qgisapp.cpp

nyalldawson committed:
New "Filled line" symbol layer type

This adds a new line symbol type which renders lines using a
fill symbol. The interior of the line is drawn using any standard
QGIS fill symbol, allowing for lines filled with gradients, line
hatches, etc.

Sponsored by North Road, thanks to SLYR

tomass committed:
Merge remote-tracking branch 'upstream/master'

nyalldawson committed:
Modernize test, make tests independent

nyalldawson committed:
[feature] Add height setting for raster fill symbol layer type

This allows users to control the width and height of the fill
pattern indepently, allowing for stretched raster fills in the
horizontal or vertical directions

Sponsored by North Road, thanks to SLYR

nyalldawson committed:
Clear some warnings

nyalldawson committed:
Regenerate expected test images for raster fills

These look different, but that's because we're now rounding
raster sizes to integer pixel sizes instead of just truncating
to integer.

nyalldawson committed:
Add tests

gacarrillor committed:
Fix typo

nyalldawson committed:
Fix documentation format of QgsCustomDropHandler

nyalldawson committed:
Update comment

nyalldawson committed:
Update date

nyalldawson committed:
Show 3d animation export progress in taskbar

nyalldawson committed:
Remove separate size unit for width and height, and just use
one unit for both to match raster marker behavior

nyalldawson committed:
Avoid deprecated method

nyalldawson committed:
Move QgsProperty::Type to Qgis, promote to enum class

NathanW2 committed:
Update missing python3-devel in INSTALL

nyalldawson committed:
Gracefully handle GLTF models where the default scene is invalid

Fixes #55182

nyalldawson committed:
Gracefully handle situation where no mesh is found in the scene

nyalldawson committed:
QgsAfsSharedData should not be a QObject

nirvn committed:
[symbology] Fix font marker inserting illegal characters in a saved project's XML document (#55041)

nyalldawson committed:
Move ::htmlMetadata from some data provider subclasses up to base class

Allows all layer types (including vector, mesh and point cloud)
providers to show custom metadata in the layer properties dialog.
Previously this was limited to raster/vector tile and tiled
scene layers only.

Also fix const for a lot of raster providers as a consequence.

nyalldawson committed:
Add missing file

raymondnijssen committed:
New styles for easter egg layers

raymondnijssen committed:
Map tips for hackfests

espinafre committed:
Looking for qwt named as qwt6, also in /usr/lib64

Adding /usr/lib64 as a possible search location for libqwt (in Gentoo,
generally 32 bit libraries reside in /usr/lib, and 64 bit, in
/usr/lib64; there might be other distributions that operate the same
way). Also, qwt might be named "libqwt6".

nyalldawson committed:
Fix a very old TODO depending on an old grass version

nirvn committed:
[authentication] Fix compilation of the AWS S3 authentication method against Qt 6.6

nirvn committed:
[wcs] Fix compilation of the WCS capabilities class against Qt 6.6

nirvn committed:
[wfs] Fix compilation of the provider library for Android against Qt 6.6

nyalldawson committed:
Fix opening settings dialog breaks oauth and ssl handling for the
rest of the session

Pretty bad bug... when you accept the settings dialog,
QgsNetworkAccessManager::setupDefaultProxyAndCache is called
to handle proxy and cache updates. This method however was also
resetting the manager's ssl error handler and auth handler to
default no-op handlers, overridding the app specific handlers
which had been setup already.

The consequence is that after accepting the options dialog, oauth
authentication and SSL error handling will be broken for the
rest of the session...

mari-s4e committed:
Update qgis-hackfests.qml with pineapples

Added pineapples on pizzas in hackfests layer

nyalldawson committed:
Traverse child nodes in gltf to vector algorithm and 2d tile rendering

jef-n committed:
[wfs] Fix compilation of the provider library with msvc (followup ff3f4660a1)

jef-n committed:
use /permissive- with msvc to get alternative operator representation
(reverts 2d4258a3)

nyalldawson committed:
If user enabled OTB plugin without setting a path, don't show
message box warnings when closing Settings dialog

This is confusing for users who have accidentally enabled the plugin.
If the path is empty, chances are the user isn't wanting to use
the OTB provider and it's nicer not to show a big confusing
message box warning to them.

pblottiere commented on src/server/qgsserverprojectutils.cpp:
```diff
@@ -195,6 +195,13 @@ bool QgsServerProjectUtils::wmsAddLegendGroupsLegendGraphic( const QgsProject &p
 
+bool QgsServerProjectUtils::wmsSkipNameForGroup( const QgsProject &project )
+{
+  const QString skipNameForGroup = project.readEntry( QStringLiteral( "WMSSkipNameForGroup" ), QStringLiteral( "/" ), "" );
```
Could you use the `readBoolEntry` method too in this case?

tomass replied:
Done

pblottiere commented on src/server/qgsserverprojectutils.h:
```diff
@@ -244,6 +244,13 @@ namespace QgsServerProjectUtils
+  /**
+   * Returns if name attribute should be skipped for groups in WMS capabilities document.
+   * \param project the QGIS project
+   * \returns if name attribute should be skipped for groups in capabilities
```
```suggestion
   * \returns if name attribute should be skipped for groups in capabilities
   * \since QGIS 3.36
```

tomass replied:
Done

pblottiere commented on src/server/services/wms/qgswmsgetcapabilities.cpp:
```diff
@@ -945,6 +945,7 @@ namespace QgsWms
 
       bool siaFormat = QgsServerProjectUtils::wmsInfoFormatSia2045( *project );
       const QStringList restrictedLayers = QgsServerProjectUtils::wmsRestrictedLayers( *project );
+      bool skipNameForGroup = QgsServerProjectUtils::wmsSkipNameForGroup( *project );
```
```suggestion
      const bool skipNameForGroup = QgsServerProjectUtils::wmsSkipNameForGroup( *project );
```

pblottiere commented:
Hello @tomass,

Thanks a lot for your work here :+1:! And sorry for the delay :/.

Can you rebase your branch and take a look to my (minor) comments please? Then it will be ready to be merged :).

nirvn committed:
[themes] Fix Night Mapping making it unjustifiably hard to identify buttons in a UI

tomass committed:
Changes according to review

tomass committed:
Changes according to review

tomass committed:
Change read to readBool

tomass committed:
Change read to readBool

tomass committed:
Merge branch 'qgis:master' into master

tomass committed:
python sip update according to .h

tomass committed:
python sip update according to .h

pblottiere commented:
Hello @tomass,

Thanks, code looks good to me! But the git history is strange (there're 68 commits), can you take a look please?

tomass commented:
> Thanks, code looks good to me! But the git history is strange (there're 68 commits), can you take a look please?
@pblottiere I think they got in when I was doing rebase. I've no idea how to remove those... any hints?

Gustry commented:
>  I was doing rebase

You have a lot of merge commit as well. I would avoid doing both(merge or rebase), choose one next time. And also avoid working on your `master` branch, you should create a new branch for a new feature.

How many commits do you have (yours) ?

tomass commented:
> You have a lot of merge commit as well. I would avoid doing both(merge or rebase), choose one next time.
> And also avoid working on your `master` branch, you should create a new branch for a new feature.
> How many commits do you have (yours) ?

I can count 12 of my commits (without merging upstream ones). I'm usually only doing merge, done rebase this time as I was asked to do that.

Gustry commented:
> I can count 12 of my commits (without merging upstream ones). I'm usually only doing merge, done rebase this time as I was asked to do that.

Rebase is better, cleaner history, but it requires a little bit more practice ;-)

Checkout a new branch based on upstream/master, with

Check your remote name with, you should have 2 remotes
```
git remote -v
```

```
git fetch upstream (if QGIS main remote is named upstream, not your own fork)
git checkout -b skip_group_wms upstream/master 
git cherry-pick ID_COMMIT_1 ID_COMMIT_2 ...  ID_COMMIT_12
git push origin skip_group_wms
```
and create a new PR

If you want, you can cherry-pick one per one, in the correct order

There are plenty of ways of fixing this, even quicker than that, but at least, I think you wouldn't loose your work (which should still be `origin/master` if something goes wrong)

tomass committed:
Merge branch 'qgis:master' into master

tomass committed:
Merge branch 'master' of https://github.com/tomass/QGIS

tomass commented:
Will create a new PR

tomass commented:
Re-created as https://github.com/qgis/QGIS/pull/55302
